subprojects/snapshots/src/main/java/org/gradle/internal/snapshot/PathUtil.java

```diff
@@ -219,51 +233,81 @@ public static int compareWithCommonPrefix(String path1, String path2, int offset
                 }
             }
         }
-        if (path1.length() == path2.length() - offset) {
+        if (relativePath.length() == absolutePath.length() - offset) {
             return accumulatedValue;
         }
-        if (path1.length() > path2.length() - offset) {
-            return isFileSeparator(path1.charAt(maxPos)) ? accumulatedValue : 1;
+        if (relativePath.length() > absolutePath.length() - offset) {
+            return isFileSeparator(relativePath.charAt(maxPos)) ? accumulatedValue : 1;
+        }
+        return isFileSeparator(absolutePath.charAt(maxPos + offset)) ? accumulatedValue : -1;
+    }
+
+    /**
+     * Checks whether the relative path given by the absolute path and the offset has the given prefix.
+     */
+    public static boolean isPrefix(String prefix, String absolutePath, int offset, CaseSensitivity caseSensitivity) {
+        int prefixLength = prefix.length();
+        if (prefixLength == 0) {
+            return true;
+        }
+        int pathLength = absolutePath.length();
+        int endOfThisSegment = prefixLength + offset;
+        if (pathLength < endOfThisSegment) {
+            return false;
+        }
+        for (int i = prefixLength - 1, j = endOfThisSegment - 1; i >= 0; i--, j--) {
+            if (!equalChars(prefix.charAt(i), absolutePath.charAt(j), caseSensitivity)) {
+                return false;
+            }
         }
-        return isFileSeparator(path2.charAt(maxPos + offset)) ? accumulatedValue : -1;
+        return endOfThisSegment == pathLength || isFileSeparator(absolutePath.charAt(endOfThisSegment));
     }
 
     /**
-     * Determines whether a path starting at an offset has the given prefix,
-     * and returns the comparison of the different prefixes if it isn't.
+     * Determines whether a relative path has the given prefix,
+     * and returns the comparison of prefix and the path if it has not.
      *
-     * Similar to {@link #isChildOfOrThis}, only that it returns whether the filePath is bigger or smaller than the current prefix.
+     * Similar to {@link #isPrefix}, only that it returns whether the path is bigger or smaller than the prefix.
+     *
+     * Examples:
+     *   some/start == some/start/subpath
+     *   some/start == some/start
+     *   some/a     < some/start
+     *   some/b     > some/a
```

Comment:
Can you add some examples for when the prefix is completely different?

Reply:
I added a sample.